Source/WebCore/css/query/MediaQueryParser.cpp

```diff
@@ -189,11 +189,9 @@ const FeatureSchema* MediaQueryParser::schemaForFeatureName(const AtomString& na
 
 void serialize(StringBuilder& builder, const MediaQueryList& list)
 {
-    for (auto& query : list) {
-        if (&query != &list.first())
-            builder.append(", "_s);
-        serialize(builder, query);
-    }
+    // Trampoline is needed to avoid lookup ambiguity.
```

Comment:
I’d like to understand more about that, and whether there are design changes we can make to make this less likely.

Reply:
Clang is unfortunately quite unhelpful with the error it gives, which is that it can't deduce "E" in interleave.

It comes down to the overload set containing multiple functions named `serialize` at the same level. I can also solve this by using a cast to the full function signature.

I think with some hand wringing we can probably construct overloads of interleave() that can do this, perhaps by having a version that looks something like:

```
template<>
decltype(auto) interleave(C&& container, (void(*)(StringBuilder&, const typename C::value_type&) each, B&& between) { ... }
```

or a variation on that using concepts, but I couldn't construct one that I was happy with.

I'll keep chipping away at it.

Reply:
I worked out a way to make this work for overloaded functions that take both a StringBuilder and value, but I still don't have a way to make it work if there are overloaded functions that take the value and return a string type.

The solution for the StringBuilder and value case was similar to what I put above. Overload interleave() with a second parameter of:

```c++
template<typename C> using EachTakingAccumulatorAndValueFunction = void(&)(StringBuilder&, const std::remove_reference_t<decltype(*std::begin(std::declval<C>()))>&);
 ```

looking like:

```c++
template<typename C, StringTypeAdaptable B> decltype(auto) interleave(const C& container, EachTakingAccumulatorAndValueFunction<C> each, B&& between)
{
    return Interleave {
        container,
        each,
        between
    };
}
```

The problem with the applying this same idea to the function returning a string value is that I don't know what the return type of that function would be and I can't figure out how to get c++ to deduce it for me.